cpp/src/arrow/testing/fixed_width_test_util.cc

```diff
@@ -0,0 +1,176 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "fixed_width_test_util.h"
+
+namespace arrow::util::internal {
+
+namespace {
+template <typename ArrowType>
+inline Status AppendNumeric(ArrayBuilder* builder, int64_t* next_value) {
+  using NumericBuilder = ::arrow::NumericBuilder<ArrowType>;
+  using value_type = typename NumericBuilder::value_type;
+  auto* numeric_builder = ::arrow::internal::checked_cast<NumericBuilder*>(builder);
+  auto cast_next_value =
+      static_cast<value_type>(*next_value % std::numeric_limits<value_type>::max());
+  RETURN_NOT_OK(numeric_builder->Append(cast_next_value));
+  *next_value += 1;
+  return Status::OK();
+}
+
+template Status AppendNumeric<Int8Type>(ArrayBuilder* builder, int64_t* next_value);
+template Status AppendNumeric<Int16Type>(ArrayBuilder* builder, int64_t* next_value);
+template Status AppendNumeric<Int32Type>(ArrayBuilder* builder, int64_t* next_value);
+template Status AppendNumeric<Int64Type>(ArrayBuilder* builder, int64_t* next_value);
```

Comment:
You don't need these (in this context) because since this template is in an anon namespace the compiler will only instantiate the 4 instances that are referred to in this compilation unit.